### PR TITLE
C++: Fix code generation for union field types that are constructed with allocators

### PR DESCRIPTION
**Pull request by jensen-reitz**
- Added `in_place_index_t` to nunavut support library; adds the resource in c++14, references `std::in_place_index_t` in c++17 builds
- Added initialization constructors to union composite type classes, takes a `nunavut::support::in_place_index_t` that specifies the alternative type to hold
- Initialization expression for `union_value` member of union composite types is now generated, with appropriate parameters, in copy and move constructors with allocators
- Replaced alternative index hardcoding in C++14 union `VariantType` internal logic with `IndexOf` property

### CODE REVIEW

File: src/nunavut/lang/cpp/templates/_fields_as_variant.j2 [hunk 1]
```diff
@@ -38,6 +51,44 @@
...
 
+        // Default constructor
+        VariantType() = default;
+
+        // Initializing constructor
+        template<std::size_t I, typename... Args, typename = std::enable_if_t<is_index_v<I>>>
+        VariantType(nunavut::support::in_place_index_t<I> i, Args&&... args) :
+            Base{std::in_place_index_t<I>{}, std::forward<Args>(args)...}
+        {
+            (void)i; // avoid unused param warning
```
Comment (thirtytwobits) [inline src/nunavut/lang/cpp/templates/_fields_as_variant.j2:62]: I think you can just omit `i` in the constructor.

**Review by thirtytwobits — COMMENTED** [review]
Looking good. Just need to add tests.

File: src/nunavut/lang/cpp/__init__.py [hunk 2]
```diff
@@ -980,6 +981,10 @@ def needs_vla_init_args(instance: pydsdl.Any, special_method: SpecialMethod) ->
     )
 
 
+def needs_variant_init_args(composite_subtype: CompositeSubType) -> bool:
```
Comment (thirtytwobits) [inline src/nunavut/lang/cpp/__init__.py:984]: docstring is missing

File: src/nunavut/lang/cpp/__init__.py [hunk 3]
```diff
@@ -994,12 +999,45 @@ def requires_initialization(instance: pydsdl.Any) -> bool:
     )
 
 
-def assemble_initializer_expression(
-    wrap: str, rhs: str, leading_args: typing.List[str], trailing_args: typing.List[str]
-) -> str:
+def prepare_initializer_args(
```
Comment (thirtytwobits) [inline src/nunavut/lang/cpp/__init__.py:1002]: docs

File: src/nunavut/lang/cpp/__init__.py [hunk 4]
```diff
@@ -994,12 +999,45 @@ def requires_initialization(instance: pydsdl.Any) -> bool:
...
+    instance: pydsdl.Any,
+    special_method: SpecialMethod,
+    composite_subtype: CompositeSubType,
+) -> typing.Tuple[typing.List[str], str, typing.List[str]]:
+    rhs: str = ""
+    leading_args: typing.List[str] = []
+    trailing_args: typing.List[str] = []
+    if needs_variant_init_args(composite_subtype):
+        leading_args.append(
+            f"nunavut::support::in_place_index_t<VariantType::IndexOf::{language.filter_id(instance)}>{{}}"
```
Comment (thirtytwobits) [inline src/nunavut/lang/cpp/__init__.py:1013]:
```suggestion
            f"{'::'.join(language.support_namespace.split('.')}::in_place_index_t<VariantType::IndexOf::{language.filter_id(instance)}>{{}}"
```

File: src/nunavut/lang/cpp/__init__.py [hunk 5]
```diff
@@ -994,12 +999,45 @@ def requires_initialization(instance: pydsdl.Any) -> bool:
...
+            trailing_args.append(constructor_args.format(MAX_SIZE=instance.data_type.capacity))
+
+    if needs_allocator(instance):
+        if language.get_option("ctor_convention") == ConstructorConvention.UsesLeadingAllocator.value:
+            leading_args.extend(["std::allocator_arg", "allocator"])
+        else:
+            trailing_args.append("allocator")
+
+    if needs_move(special_method):
+        rhs = "std::move({})".format(rhs)
```
Comment (thirtytwobits) [inline src/nunavut/lang/cpp/__init__.py:1034]: use f-string here

File: src/nunavut/lang/cpp/__init__.py [hunk 6]
```diff
@@ -994,12 +999,45 @@ def requires_initialization(instance: pydsdl.Any) -> bool:
...
+    trailing_args: typing.List[str] = []
+    if needs_variant_init_args(composite_subtype):
+        leading_args.append(
+            f"nunavut::support::in_place_index_t<VariantType::IndexOf::{language.filter_id(instance)}>{{}}"
+        )
+
+    if needs_initializing_value(special_method):
+        instance_id = language.filter_id(instance)
+        if needs_rhs(special_method):
+            rhs = "rhs."
```
Comment (thirtytwobits) [inline src/nunavut/lang/cpp/__init__.py:1019]: "rhs" should be a token retrieved from properties.yaml. Can you add a key for this and replace all hard-coded values?

File: src/nunavut/lang/cpp/__init__.py [hunk 7]
```diff
@@ -994,12 +999,45 @@ def requires_initialization(instance: pydsdl.Any) -> bool:
...
+        else:
+            trailing_args.append("allocator")
+
+    if needs_move(special_method):
+        rhs = "std::move({})".format(rhs)
+
+    return (leading_args, rhs, trailing_args)
+
+
+def assemble_initializer_expression(rhs: str, leading_args: typing.List[str], trailing_args: typing.List[str]) -> str:
```
Comment (thirtytwobits) [inline src/nunavut/lang/cpp/__init__.py:1039]: should this be `_private`?

File: src/nunavut/lang/cpp/__init__.py [hunk 8]
```diff
@@ -1008,40 +1046,34 @@ def assemble_initializer_expression(
...
 
-        if needs_move(special_method):
-            wrap = "std::move"
+    return value_initializer
 
-        value_initializer = assemble_initializer_expression(wrap, rhs, leading_args, trailing_args)
 
-    return value_initializer
+@template_language_test(__name__)
+def filter_needs_allocator(language: Language, instance: pydsdl.Any) -> bool:
```
Comment (thirtytwobits) [inline src/nunavut/lang/cpp/__init__.py:1074]:
```suggestion
def filter_needs_allocator(_: Language, instance: pydsdl.Any) -> bool:
```

File: src/nunavut/lang/cpp/__init__.py [hunk 9]
```diff
@@ -994,12 +999,45 @@ def requires_initialization(instance: pydsdl.Any) -> bool:
...
+def prepare_initializer_args(
+    language: Language,
+    instance: pydsdl.Any,
+    special_method: SpecialMethod,
+    composite_subtype: CompositeSubType,
+) -> typing.Tuple[typing.List[str], str, typing.List[str]]:
+    rhs: str = ""
+    leading_args: typing.List[str] = []
+    trailing_args: typing.List[str] = []
+    if needs_variant_init_args(composite_subtype):
```
Comment (thirtytwobits) [inline src/nunavut/lang/cpp/__init__.py:1011]: false branch not covered in python test

File: src/nunavut/lang/cpp/__init__.py [hunk 10]
```diff
@@ -994,12 +999,45 @@ def requires_initialization(instance: pydsdl.Any) -> bool:
...
+    if needs_vla_init_args(instance, special_method):
+        constructor_args = language.get_option("variable_array_type_constructor_args")
+        if isinstance(constructor_args, str) and len(constructor_args) > 0:
+            trailing_args.append(constructor_args.format(MAX_SIZE=instance.data_type.capacity))
+
+    if needs_allocator(instance):
+        if language.get_option("ctor_convention") == ConstructorConvention.UsesLeadingAllocator.value:
+            leading_args.extend(["std::allocator_arg", "allocator"])
+        else:
+            trailing_args.append("allocator")
```
Comment (thirtytwobits) [inline src/nunavut/lang/cpp/__init__.py:1031]: not covered in python test

File: src/nunavut/lang/cpp/__init__.py [hunk 11]
```diff
@@ -994,12 +999,45 @@ def requires_initialization(instance: pydsdl.Any) -> bool:
...
+
+    if needs_initializing_value(special_method):
+        instance_id = language.filter_id(instance)
+        if needs_rhs(special_method):
+            rhs = "rhs."
+        rhs += f"get_{instance_id}()" if composite_subtype is CompositeSubType.Union else instance_id
+
+    if needs_vla_init_args(instance, special_method):
+        constructor_args = language.get_option("variable_array_type_constructor_args")
+        if isinstance(constructor_args, str) and len(constructor_args) > 0:
```
Comment (thirtytwobits) [inline src/nunavut/lang/cpp/__init__.py:1024]: false branch not covered in python test.

File: src/nunavut/lang/cpp/__init__.py [hunk 12]
```diff
@@ -1008,40 +1046,34 @@ def assemble_initializer_expression(
...
+        value_initializer = assemble_initializer_expression(rhs, leading_args, trailing_args)
 
-        if needs_move(special_method):
-            wrap = "std::move"
+    return value_initializer
 
-        value_initializer = assemble_initializer_expression(wrap, rhs, leading_args, trailing_args)
 
-    return value_initializer
+@template_language_test(__name__)
```
Comment (thirtytwobits) [inline src/nunavut/lang/cpp/__init__.py:1073]: this filter not coverend in python tests.

File: verification/cpp/suite/test_unionant.cpp [hunk 13]
```diff
@@ -66,22 +66,69 @@ TEST(UnionantTests, get_set_lvalue)
...
+    const uavcan::primitive::array::Integer32_1_0* p_int32 =
         uavcan::_register::Value_1_0::VariantType::get_if<ValueType::VariantType::IndexOf::integer32>(&a.union_value);
-    ASSERT_EQ(nullptr, p);
+
+    ASSERT_NE(nullptr, p_empty);
+    ASSERT_EQ(nullptr, p_int32);
 }
 
+/**
+ *  Verify that the variant value can be fetched only at the alternative index for the type being held
```
Comment (thirtytwobits) [inline verification/cpp/suite/test_unionant.cpp:87]: Missing a test that actually passes an allocator through to a variant type constructor.

**Review by thirtytwobits — CHANGES_REQUESTED** [review]

File: src/nunavut/lang/cpp/__init__.py (hunk 7 above)
Comment (jensen-reitz) [reply src/nunavut/lang/cpp/__init__.py:1039]: This and `assemble_initializer_args()` can be made private

**Comment by thirtytwobits** [comment]
Closing as abandoned.

**Pull request closed without merging**